Introduce target_stack property to update specific stacks and enable finder preview

minhngocd opened the pull request:
Introduced `target_stack` property to replace pre_production flag
    
The new `target_stack` property will publish anything with a target stack of `live` to the live stack and draft stack and anything else to the draft stack only.
This allows us to surface appropriate preview links and live links for users later.

Delete anything associated with pre_production and updated documentation

https://trello.com/c/JrkAwfDU/2786-dynamically-update-stacks-based-on-state-of-specialist-finder    

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.

lauraghiorghisor-tw committed:
Add logs to distinguish which publishing api requests are being made

lauraghiorghisor-tw committed:
Remove usages of pre_production flag

minhngocd committed:
Documentation update for target_stack property

Deployment flow for new finder updated to take into account finder previews.
Updated the other amending facet and values documentation to inline specific instruction to not get confused with the new finder deployment flow.

Co-authored-by: Laura Ghiorghisor <laura.ghiorghisor@digital.cabinet-office.gov.uk>

ChrisBAshton commented on lib/documents/schemas/aaib_reports.json:
```diff
@@ -1,4 +1,5 @@
 {
+  "target_stack": "live",
```
To make it easier to review, I'd move the config update into its own commit (i.e. add all the `target_stack` properties in one commit, then the logic that uses it would be added in the next commit).

ChrisBAshton commented on lib/documents/schemas/veterans_support_organisations.json:
```diff
@@ -1,5 +1,5 @@
 {
-  "pre-production": true,
```
I'd also remove the pre-production config and logic in a separate commit, once the `target_stack` stuff has been implemented.

EDIT: looks like you've already tried to do this - in which case, just double-check this and rebase it into the (currently) third commit.

ChrisBAshton commented on docs/creating-editing-and-removing-specialist-document-types-and-finders.md:
```diff
@@ -72,17 +72,24 @@ If your email sign up page should have checkboxes (e.g. [cma-cases](https://www.
 
 ## 5. Deploy and publish <a id="deploy-and-publish"></a>
```
Is it worth splitting this out into:

```
## 5. Deploy a finder to the draft stack (for previewing)

## 6. Publish a finder
```

ChrisBAshton reviewed (commented):
LGTM 👍 a few comments below.

minhngocd committed:
Introduced new target_stack property to document schemas

This will enable us to publish to draft stack for previews of the finder. Logic to be implemented in the next commit.

Co-authored-by: Alex Newton <alex.newton@digital.cabinet-office.gov.uk>
Co-authored-by: Laura Ghiorghisor <laura.ghiorghisor@digital.cabinet-office.gov.uk>

minhngocd committed:
Implement logic to only publish finders with target_stack of live

The new `target_stack` property will publish anything with a target stack of `live` to the live stack and draft stack and anything else to the draft stack only.
This allows us to surface appropriate preview links and live links for users later.

Co-authored-by: Alex Newton <alex.newton@digital.cabinet-office.gov.uk>
Co-authored-by: Laura Ghiorghisor <laura.ghiorghisor@digital.cabinet-office.gov.uk>

ChrisBAshton commented on lib/documents/schemas/veterans_support_organisations.json:
```diff
@@ -1,4 +1,5 @@
 {
+  "target_stack": "draft",
```
This should probably be `live`, unless there are any config changes here that aren't on the [live site](https://www.gov.uk/support-for-veterans) yet.

minhngocd replied:
oooh thanks for catching this! i lost it amongst the merge conflicts

ChrisBAshton commented on lib/tasks/publishing_api.rake:
```diff
@@ -12,19 +12,6 @@ namespace :publishing_api do
     end
   end
 
-  desc "Publish pre-production finders only, to the Publishing API"
```
This removal should probably be fixed up to the "Remove usages of pre_production flag" commit.

minhngocd replied:
But the logic is switched and changed in this commit, so this rake task would be broken anyways - we thought it'd be best to remove in this commit

ChrisBAshton commented on spec/lib/publishing_api_finder_publisher_spec.rb:
```diff
@@ -109,12 +106,8 @@ def make_finder(base_path, overrides = {})
       end
     end
 
-    context "when the finder isn't `pre_production`" do
```
Ideally this (and the removals below) should also be in the "Remove usages of pre_production flag" commit, but I appreciate it may be hard to unpick as at this point we've changed the deployment logic 🤔

minhngocd replied:
Similarly here - if we leave these tests here they would be failing

ChrisBAshton commented on docs/creating-editing-and-removing-specialist-document-types-and-finders.md:
```diff
@@ -36,7 +36,7 @@ When the PR is reviewed and its tests passing, it can be merged and deployed at
 See [CMA cases](https://github.com/alphagov/specialist-publisher/blob/main/lib/documents/schemas/cma_cases.json).
 
-New formats are often requested to be deployed in "pre-production mode", which is configured in this step ([example](https://github.com/alphagov/specialist-publisher/blob/f8e93142dfad6f3971a73c923b01f2e7352bdb54/lib/documents/schemas/tax_tribunal_decisions.json#L64)). `pre-production` documents are only publishable on development and integration.
+New formats should be added with `state: "draft"` so that departments can preview the changes.
```
```suggestion
New formats should be added with `state: "draft"` so that departments can preview the finder before you publish it.
```

ChrisBAshton commented on docs/creating-editing-and-removing-specialist-document-types-and-finders.md:
```diff
@@ -70,20 +70,29 @@ The email sign up page is rendered by [Finder Frontend](https://github.com/alpha
       - Ensure you deploy Publishing API first, to avoid schema validation errors.
       - Also deploy Email Alert API if you have made changes to it.
-   2. [Reindex the govuk Elasticsearch index](https://docs.publishing.service.gov.uk/manual/reindex-elasticsearch.html#how-to-reindex-an-elasticsearch-index).
+   1. [Reindex the govuk Elasticsearch index](https://docs.publishing.service.gov.uk/manual/reindex-elasticsearch.html#how-to-reindex-an-elasticsearch-index).
```
Hmmm - given it's only the finder 'skeleton' that's previewable, I'm not sure there's much point to the indexing here. Should this step (and its substeps) be moved into the publish section instead?

ChrisBAshton reviewed (commented):
Thanks @minhngocd - looking good, just a few follow-up comments below.

ChrisBAshton reviewed (approved):
Nice job 👏
